pkg/gatewayserver/gatewayserver.go

```diff
@@ -752,11 +728,25 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 		host := &upstreamHost{
 			name:          name,
 			handler:       handler,
-			handleCh:      make(chan upstreamItem),
+			gtw:           gtw,
 			correlationID: fmt.Sprintf("gs:up:host:%s", events.NewCorrelationID()),
 		}
+		wp, err := workerpool.NewWorkerPool(workerpool.Config{
+			Component:     gs,
+			Context:       ctx,
+			Name:          fmt.Sprintf("upstream_handlers_%v", protocol),
+			CreateHandler: workerpool.StaticHandlerFactory(host.handlePacket),
+			MinWorkers:    -1,
+			MaxWorkers:    32,
+			QueueSize:     -1,
+		})
+		if err != nil {
+			conn.Disconnect(err)
+			return
+		}
+		defer wp.Wait()
+		host.pool = wp
 		hosts = append(hosts, host)
-		defer host.handleWg.Wait()
```

Comment:
This should've deadlocked months ago when we introduced TxAcks, since the tests didn't actually read the NS channel, only the events.

I don't see why this `Wait()` didn't deadlock before, this is weird.